feature(backend): Enabling Analyzer Configuration

xoscar opened the pull request:
This PR updates the architecture for plugins and rules, keeping configuring plugins more straightforward and flexible.

## Changes

- Decouples rules from plugins and allows for a configuration-based setup
- Moves the `linter` out of the `models` to the `linter` directory
- Updates plugins and rules interfaces to support an analyzer configuration
- Moves metadata away from linter execution, leaving it to be config only
- Enables configuration for `error level` and `weight` for rules

## Fixes

- #2773 

## Checklist

- [x] tested locally
- [ ] added new dependencies
- [ ] updated the docs
- [ ] added a test

Analyzer Arch Changes
https://www.loom.com/share/03e95872d383496aa81591be9d772123

Features & bug fixes
https://www.loom.com/share/0517b986f19941f3b41c5187d2344500

xoscar commented on server/executor/linter_runner.go:
```diff
@@ -100,12 +100,11 @@ func (e *defaultlinterRunner) startWorker() {
-	linter := linter.NewLinter(lintResource, linter.AvailablePlugins...)
 
-	shouldSkip, reason := linter.ShouldSkip()
+	shouldSkip := lintResource.ShouldSkip()
```
The analyzer resource should make the decisions of running or not the linter as it contains the configuration

xoscar commented on server/linter/analyzer/analyzer_entities.go:
```diff
@@ -28,9 +30,25 @@ type (
+		Description string `json:"-"`
+	}
+
+	LinterRule struct {
```
This a new block, allowing users to configure rules

xoscar commented on server/linter/analyzer/analyzer_entities.go:
```diff
@@ -40,18 +58,130 @@ func (l Linter) Validate() error {
 	for _, p := range l.Plugins {
-		if p.Name == "" {
-			return fmt.Errorf("plugin name cannot be empty")
+		plugin, ok := findPlugin(p.Id, DefaultPlugins)
```
Validates that the plugins and rules coming from the new resource instance match what we have in the system, this should be more flexible in the future when we implement custom analyzer profiles

xoscar commented on server/linter/analyzer/analyzer_entities.go:
```diff
@@ -40,18 +58,130 @@ func (l Linter) Validate() error {
+	return !l.Enabled
+}
+
+func (l Linter) WithMetadata() (Linter, error) {
```
Combines the analyzer configuration (Database, CLI, FE) with the metadata stored within the repo, as it is not available for users to update (today)

xoscar commented on server/linter/analyzer/analyzer_metadata.go:
```diff
@@ -0,0 +1,139 @@
+
+var (
+	// plugins
+	StandardsId = "standards"
```
Names, ids, descriptions, etc

xoscar commented on server/linter/linter.go:
```diff
@@ -2,107 +2,81 @@ package linter
-		linter_plugin_security.NewSecurityPlugin(),
-		linter_plugin_common.NewCommonPlugin(),
-	}
+	commonPlugin = plugins.NewPlugin(
```
Static registration of the linter configuration, we do not have much flexibility today, but in the future, the linter runner could generate a dynamic plugin registry with custom rules and plugins defined by user profile, test or test run levels

xoscar commented on server/linter/plugins/plugins_entities.go:
```diff
@@ -0,0 +1,62 @@
+	return p.ruleRegistry
+}
+
+func (p BasePlugin) Execute(ctx context.Context, trace model.Trace, config analyzer.LinterPlugin) (analyzer.PluginResult, error) {
```
All current plugins have the same execution logic, so instead of having to create a new `plugin` file for each, we use the default, and if needed, we can create specific plugin files implementing the `Plugin` interface to override execute method with custom logic

xoscar commented on server/linter/rules/rules_entities.go:
```diff
@@ -0,0 +1,27 @@
+	return r.id
+}
+
+func (r BaseRule) Evaluate(ctx context.Context, trace model.Trace, config analyzer.LinterRule) (analyzer.RuleResult, error) {
```
Rules on the other hand, have very custom logic, which makes it more difficult to encapsulate in a single default method, for these, we should create an implementation of the `Rule` interface per rule

schoren replied:
I would say there's no need for this `BaseRule`. it can be replaced by a `func (r MyRule) ID() string {return someconstantID}` func, satisfying the interface. This also makes the ID immutable, and that's good in this context

xoscar commented on server/resourcemanager/resource_manager.go:
```diff
@@ -222,6 +222,15 @@ func (m *manager[T]) create(w http.ResponseWriter, r *http.Request) {
 		targetResource.Spec = m.rh.SetID(targetResource.Spec, m.config.idgen())
 	}
 
+	if err := targetResource.Spec.Validate(); err != nil {
```
I added this validate step to the resource manager, as it wasn't currently in use, not sure if I should better move this to another PR

danielbdias reviewed (approved):
The code overall is ok! Good job!

One thing to consider in the future is how we can test these rules to guarantee that they are working as intended and are only sensitive to changes in the future without us knowing it.

xoscar commented:
> The code overall is ok! Good job!
> 
> One thing to consider in the future is how we can test these rules to guarantee that they are working as intended and are only sensitive to changes in the future without us knowing it.

Hey @danielbdias thanks for the review, I have added tests for the analyzer and linter critical points, and will add more for rules in the following PRs!

schoren commented on server/linter/rules/not_empty_attribute_rule.go:
```diff
@@ -0,0 +1,56 @@
+
+	if config.ErrorLevel != analyzer.ErrorLevelDisabled {
+		for _, span := range trace.Flat {
+			emptyAttributes := make([]string, 0)
+			for name, value := range span.Attributes {
+				if value == "" {
+					emptyAttributes = append(emptyAttributes, name)
+				}
+			}
+
+			errors := make([]analyzer.Error, 0, len(emptyAttributes))
+			for _, emptyAttribute := range emptyAttributes {
+				errors = append(errors, analyzer.Error{
+					Value:       emptyAttribute,
+					Description: fmt.Sprintf(`Attribute "%s" is empty`, emptyAttribute),
+				},
+				)
+			}
+
+			if len(errors) > 0 {
+				passed = false
+			}
+
+			res = append(res, analyzer.Result{
+				SpanID: span.ID.String(),
+				Passed: len(emptyAttributes) == 0,
+				Errors: errors,
+			})
+		}
+	}
```
Looks like there's a dobule loop that could be avoided here
```suggestion
			errors := make([]analyzer.Error, 0, len(emptyAttributes))
			for name, value := range span.Attributes {
				if value == "" {
					errors = append(errors, analyzer.Error{
						Value:       emptyAttribute,
						Description: fmt.Sprintf(`Attribute "%s" is empty`, emptyAttribute),
					})
				}
			}

			if len(errors) > 0 {
				passed = false
			}

			res = append(res, analyzer.Result{
				SpanID: span.ID.String(),
				Passed: passed,
				Errors: errors,
			})
		}
	}
```

schoren commented on server/linter/rules/required_attribute_rule.go:
```diff
@@ -0,0 +1,37 @@
+	var allPassed bool = true
+
+	if config.ErrorLevel != analyzer.ErrorLevelDisabled {
+		for _, span := range trace.Flat {
+			res = append(res, r.validateSpan(span))
+		}
+
+		for _, result := range res {
+			if !result.Passed {
+				allPassed = false
+			}
+		}
```
same here

```suggestion
		for _, span := range trace.Flat {
			analyzerResult := r.validateSpan(span)
			res = append(res, analyzerResult)
			if !analyzerResult.Passed {
				allPassed = false
			}

		}
```

schoren commented on server/linter/rules/rules_registry.go:
```diff
@@ -0,0 +1,52 @@
+	Evaluate(context.Context, model.Trace, analyzer.LinterRule) (analyzer.RuleResult, error)
+}
+
+type RuleRegistry interface {
```
Who depends on this interface? can the dependees directly reference a struct? What benefits do we get from this interface?

If a client depends on this interface, it will be requiring this package. Given that we only support one implementation of the `RuleRegistry`, and it's in memory without any dependencies, doesn't look like we would need to mock it, for example, or replace it by another implementation.

xoscar replied:
Makes sense, I think the best case is just to remove the interface

schoren commented on server/resourcemanager/resource_manager.go:
```diff
@@ -255,6 +255,15 @@ func (m *manager[T]) doCreate(w http.ResponseWriter, r *http.Request, encoder En
 		specs = m.rh.SetID(specs, m.config.idgen())
 	}
 
+	if err := specs.Validate(); err != nil {
```
niceeeee

schoren commented on server/resourcemanager/resource_manager.go:
```diff
@@ -288,6 +297,15 @@ func (m *manager[T]) upsert(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	if err := targetResource.Spec.Validate(); err != nil {
```
this is already handled on `doCreate`, and if you move [this](https://github.com/kubeshop/tracetest/pull/2842/files#diff-73567862de3c0b65102c45d3df1f98a49f2649fe588140f55a56fa693068bf51R351) to happen within the `doUpdate` method, we can have both cases covered and remove the need for extra work here

schoren reviewed (changes requested):
It looks good in general. I think we should try to embrace go's naming conventions, so it'd be great to rename all the `Id`s to `ID`.

I outlined a couple of duplicated loops and bugs, not necessarily related specifically to this work, but since we're here 🤷 

Finally, more oriented to the architecutre:
1. I would remove the `BaseRule`, it's being used kind of as a base class, but it doesn't make a lot of sense here IMO.
2. We should try to reduce the amount of interfaces that are not strictly required. "Accept interfaces, return structs". You can leverage go's duck typing and declare interfaces where you want to avoid direct references to a package.
3.

xoscar commented:
Hey @schoren thanks for the review it was really insightful! I think I addressed all of your comments and updated the code accordingly, let me know if there is something else I could do

schoren reviewed (approved):
Thanks @xoscar! this looks very consistent now. I would strongly suggest to remove the remaining duplicated loops ([here](https://github.com/kubeshop/tracetest/pull/2842/files/e7e3d02e5effa551f21fd12021d9edae10ca7ede#r1252418244) and [here](https://github.com/kubeshop/tracetest/pull/2842/files/e7e3d02e5effa551f21fd12021d9edae10ca7ede#r1252418797)) and remove the `BaseRule`, since it's now unused (I think?)

Other than that, LGTM
